### ISSUE TITLE
Annotate Images

### ISSUE THREAD

**Issue by jasondavis** [twostairs/paperwork#235]
I forgot about this fairly new feature in Evernote.  You can right click on an image in a Note and click on "Annotate Image" which then opens the image in an editor with tools to annotate with text overlay, lines, arrows, etc...

This might be a cool feature to add to Paperwork if we can find a nice open source editor/annotator type script to use?  If we are adding OCR support than I feel this feature might not be such a long stretch either and especially if we can find a library that does the bulk of the work!

Would love to hear others thoughts on:

1) IF you would like this feature and find it usable?

2) Have any suggestions or ideas on the best method to implement it?

3) Any other comments on this feature?

![2015-02-10_21-02-55](https://cloud.githubusercontent.com/assets/186818/6140609/699bc79c-b168-11e4-961f-9fac2e195b32.png)

**Comment by jasondavis** [comment]
In my research so far, I have not been able to find any good open source apps that we could use for this =(

The closest thing I have found is an Annotation script that I used on a previous project and it is very basic, it only allows you to insert an arrow, or some other type image onto your image and it puts it overlayed on the image anywhere you click.  It simply saves the coordinates of where you had clicked and position a small arrow type image at that location on top of your big image.

The image below is an example of how I used it in another project where the goal was for people that had to return a broken neon sign.  It allowed them to upload an image of the broken sign and then the Annotation script would let them click anywhere on the image and it would insert the arrow image where they clicked.  You can see for my purpose, it worked great!  However this is about the best I can find so far for an open source annotation JavaScript library.  

If anyone has ideas, please do share!

![4_step](https://cloud.githubusercontent.com/assets/186818/6202422/f624fd1c-b4b7-11e4-8731-c649f2053977.png)

**Comment by jasondavis** [comment]
I think I am looking at this from the wrong angle...

What I should be looking for is not a simple open source Annotation script, but instead a simple editor.  There are tons of editors that allow lines, arrows, text, shapes, etc... Either way if anyone has ideas for a good library or way to do this, share here for all of us to see please!

**Comment by ntnn** [comment]
Imho the easiest way would be to use a simple annotation script that allows to add text/arrows/... to an image and then save it as a new version of the attachment - this, of course, would require to also version control versions of an attachment, which currently doesn't seem to be the case (but a dev would need so say something about that).

That way would also be easier to implement for other applications that access paperwork through the API.

**Comment by devilx** [comment]
@ntnn correct. Attachments aren't being version controlled. But a buddy of mine, @frontend-gmbh, built a nice preview of how this could be done using pure JS/CSS. The arrows and other annotations there are being saved within an array, which could easily be added to the attachment, using a new database-column. Then these annotations would remain dynamic / changeable. When the user want's to send the annotated image via mail (or however), some exporting to PNG would be needed, though.

Maybe I can convince @frontend-gmbh in working on this issue... :-)

**Comment by ntnn** [comment]
@devilx Personally I'd step back from that idea.

In the view from an API dev this would need three additional steps:
Create a new version of the note through updating, delete the old attachment from the new version, upload the 'new' attachment.

Wouldn't it be easier to make attachments version controlled? That would also enable users to use one attachment for multiple notes (this would actually making attachments similar to tags, thinking about it).  
For pictures this can, of course, cost quite a bit of space if it is edited often.

**Comment by devilx** [comment]
@ntnn well, we have two different ballparks here:

1) The way annotations are being saved:

I my opinion, we should never *ever* modify a user-uploaded document/image/whatever. Therefor I pledge for modifying not the source image, nor saving a modified version as whatever kind of file (either a copy or a versioned successor), but rather saving that special metadata somewhere and recall it for displaying and rendering purpose.

2) Whether we should make attachments versioned or not:

Imagine the complexity of having notes, which are versioned and also attaching a set of versioned attachments to that notes. Where would you attach them anyway? Currently, when a new note version was created, all attachments from the previous version are also attached to the new version. Now, if we continue like that, we would need an attachment model that is attached to every new version, which for its part also has sub-versions, that are attached to it.

I'm not sure, whether that's a good idea in terms of complexity. I support your argument though, that from the API point of view, copying attachments just isn't the way to go either.

**Comment by jasondavis** [comment]
@devilx As for keeping any annotation data separate from the file and not like saved into an actual image file I agree 100% and wanted to add that Evernote does not save the image either!

I see this might be a difficult feature but still worth having on the roadmap for someday =)

**Comment by ntnn** [comment]
@devilx 

>1) The way annotations are being saved:

>I my opinion, we should never ever modify a user-uploaded document/image/whatever. Therefor I pledge for modifying not the source image, nor saving a modified version as whatever kind of file (either a copy or a versioned successor), but rather saving that special metadata somewhere and recall it for displaying and rendering purpose.

What I can understand - but this will create problems with third-party applications such as desktop clients or mobile apps. All these applications would need to write their own interpretation of the way the metadata is saved and applied (if there isn't such a project yet).
I'd prefer separate meta-data too, but I see too many problems with that.

>2) Whether we should make attachments versioned or not:

>Imagine the complexity of having notes, which are versioned and also attaching a set of versioned attachments to that notes. Where would you attach them anyway? Currently, when a new note version was created, all attachments from the previous version are also attached to the new version. Now, if we continue like that, we would need an attachment model that is attached to every new version, which for its part also has sub-versions, that are attached to it.

I tried to visualize my idea how it should be set up with this uml: 
![paperwork 1](https://cloud.githubusercontent.com/assets/10514301/6208181/02d5942a-b5b4-11e4-9f25-8d518d8e85c8.jpg)



AttachmentVersion then should either hold the metadata and Attachment holds the file itself or AttachmentVersion holds the different Versions of files.

**Comment by devilx** [comment]
@ntnn regarding 2: I understood you quite well initially. I will try to explain the fundamental issue here using a more precise scenario:

You create a new note. It contains one ```version-1```. You change the note's text and save it. ```version-2``` is being created. Now you add an attachment to that note. ```version-3``` is being created, towards which ```attachment-1``` is being linked, which for its part contains an ```attachmentversion-1```. You change that attachment, meaning that ```attachmentversion-2``` is being created. You change the note text again, meaning, that now ```version-4``` is being created. Now, what happens to the attachments? If you want to do it right, you would need to duplicate ```attachment-1```, with the copy containing only the last ```attachmentversion-``` form ```attachment-1```. By that, when the user utilizes the wayback machine, he would be able to skim through the version and the attachment versions for every version. If you'd only just move the attachment to every new version (as we do it right now, with the un-versioned attachments), the user would need a separate wayback machine for each attachment, in addition to the wayback machine on top of the note that lets him skim through the note versions. Because you couldn't tell what attachmentversion was active for the note version 46, for example.

This whole version would end up in a huge piece of code that basically no one would ever really benefit from. People who need things so heavily versioned probably use GIT/SVN/etc. anyway.

Regarding 1: Yup, that's totally right. We would need every platform to interpret the annotation meta in the same way, in order to be able to display it. That is a challenger, but none that's impossible to do. For example, if we would use some sort of SVG format for drawing the shapes/annotations, we would basically only need to use existing SVG implementations for the native clients on every platform. An annotation would then be nothing more than an SVG "image", that has our user's upload (as PNG or whatever) set as background and where the user drew all kind of shapes on it. We could even export the whole thing as SVG, if the user would like to.

**Comment by ntnn** [comment]
@devilx

I see the problem you mean now - but that could be solved by directly linking a `noteversion` to an `attachmentversion`. But still that would use up a lot of space and additionally it would need to check for remaining version-links when removing a note or attachment, etc pp.

All in all your idea is better - thanks for explaining!
